Bring test coverage above 95%

ShogunPanda opened the pull request:
This PR provides mainly two things:

1. Make sure it is possible to connect to IPv6 IPs. @mcollina @ronag Please double check this. I randomly found this while bringing coverage up.
2. Bring coverage above 95% (now it is around 98%).

mcollina commented on index.js:
```diff
@@ -32,7 +32,7 @@ function dispatchFromAgent (fn) {
 
     const { origin, pathname, search } = util.parseURL(url)
-    const path = `${pathname || '/'}${search || ''}`
+    const path = `${pathname /* istanbul ignore next */ || '/'}${search || ''}`
```
why ignoring? Could this even happen?

ronag replied:
This can't happen. Just remove the `||`.

ShogunPanda replied:
Done!

mcollina reviewed (commented):
Can you please add the -100 check of tap so we do not regress?

ronag commented on lib/agent/redirect.js:
```diff
@@ -16,33 +13,9 @@ class RedirectHandler {
-    this.handler.onUpgrade(statusCode, headers, socket)
+    this.abort = abort
+    this.handler.onConnect(abort)
   }
```
All if these can happen. Please put them back.

ShogunPanda replied:
Done. I was not able to find in the code how to reach those hooks so I leave it up to you.

ronag commented on lib/agent/redirect.js:
```diff
@@ -61,7 +34,7 @@ class RedirectHandler {
 
     const { origin, pathname, search } = util.parseURL(new URL(this.location, this.opts.origin))
-    const path = search ? `${pathname || '/'}${search || ''}` : pathname
+    const path = search ? `${pathname /* istanbul ignore next */ || '/'}${search}` : pathname
```
Just remove the `||`.

ShogunPanda replied:
Done!

ronag commented on lib/agent/redirect.js:
```diff
@@ -83,16 +56,38 @@ class RedirectHandler {
 
   onComplete (trailers) {
     if (this.location) {
-      assert(!trailers || trailers.length === 0, 'redirect should not have trailers')
```
This can happen. Please put back.

ShogunPanda replied:
I'm not sure I understand.
In the comment I'm not saying it cannot happen, I'm saying that we don't handle them, like we do with the body.

I just forgot to remove the `this.agent.dispatch` line below. Do you agree or do you explicitily want to handle trailers?

ronag replied:
I would add a comment so that it's explicit that we ignore them on purpose

ShogunPanda replied:
Ok, and then I'll redirect to the full explanation above. Understood. Updated coming soon.

ShogunPanda replied:
Done!

ronag commented on lib/client-pool.js:
```diff
@@ -104,6 +104,7 @@ class Pool extends EventEmitter {
 
     if (this[kConnections] && this[kClients].length === this[kConnections]) {
       for (const { busy } of this[kClients]) {
+        /* istanbul ignore next */
```
This should be possible to happen. Please remove ignore.

ShogunPanda replied:
Done. I leave it to you to add a test for it. I've tried with no luck.

ronag replied:
Please remove the ignore

ShogunPanda replied:
Update coming soon!

ShogunPanda replied:
Done!

ronag commented on lib/core/client.js:
```diff
@@ -65,7 +65,7 @@ function getServerName (client, host) {
     util.getServerName(host) ||
     (client[kTLSOpts] && client[kTLSOpts].servername) ||
-    util.getServerName(client[kUrl].host || client[kUrl].hostname) ||
+    util.getServerName(client[kUrl].host /* istanbul ignore next */ || client[kUrl].hostname) ||
```
This can happen. Please remove ignore.

ShogunPanda replied:
How comes? Since we are on Node LTS (so > 10 at the moment), `client[kUrl]` is a instance of `URL` (the WHATWG one), so if `host` is not defined, neither is `hostname`.

Am I wrong?

ronag replied:
you are right, then please remove the `||` part.

ShogunPanda replied:
Update coming soon!

ShogunPanda replied:
Done!

ronag reviewed (changes requested):
Excellent work. Please fix the comments and remove all the istanbul ignores I think some of them we can make tests for. I'd like to have a shot at them as well later.

ShogunPanda commented:
@mcollina I'm not sure I can at the moment. Since Robert has to provide missing tests for the cases I unignored, if I do that now the CI will fail. Unless we can agree for a lower threshold, like 95%.

@ronag I've made most of the fixes you asked for, except for two I left comment for. Do you want me to remove ALL remaining istanbul ignores? Just double checking.

ronag commented on lib/agent/index.js:
```diff
@@ -94,10 +94,10 @@ class Agent extends EventEmitter {
       return client.dispatch(opts, handler)
     }
 
+    /* istanbul ignore next */
```
This can happen? It's just the data event that we don't bother testing for.

ShogunPanda replied:
If you look at line 90, there is an entire check about body  empty stream. So, the if on 98 is only satisfied by 
I was not able to simulate a empty stream which emits so I skipped the if check as well.

ronag replied:
I guess it doesn't matter so much.

ronag reviewed (changes requested):
Please remove all the ignores. I'll have a go at it later.

ronag commented on lib/core/client.js:
```diff
@@ -863,7 +870,18 @@ function onSocketSession (session) {
+  // Resolve ipv6
+  if (hostname.startsWith('[')) {
+    const idx = hostname.indexOf(']')
+    const ip = idx !== -1 ? hostname.substr(1, idx - 1) /* istanbul ignore next */ : hostname
```
remove ternary and just make an `assert(idx !== -1)`

ShogunPanda replied:
It should not because URL parsing would fail.
Ok, will do that!

ronag replied:
I'm fine with doing istanbul ignore or assertions.

ShogunPanda replied:
Done!

ronag commented on lib/core/client.js:
```diff
@@ -863,7 +870,18 @@ function onSocketSession (session) {
+    /* istanbul ignore else */
+    if (net.isIP(ip)) {
+      hostname = ip
+    }
```
is the else possible here? what happens then?

ShogunPanda replied:
Same thing. I think I will add an assert here as well. Else should not be possible.

ShogunPanda replied:
Done!

ronag commented on lib/core/util.js:
```diff
@@ -167,7 +167,7 @@ module.exports = {
   bodyLength,
   isBuffer,
-  queueMicrotask: global.queueMicrotask || (cb => Promise.resolve()
+  queueMicrotask: global.queueMicrotask /* istanbul ignore next */ || (cb => Promise.resolve()
```
Could we just do `global.queueMicrotask = null` in a test?

ShogunPanda replied:
Mh...we could but I definitely don't like playing with global.
Also, since we only support Node and we are on >10, do we really need this?

mcollina replied:
I would remove this completely. `queueMicrotask` should be available in 12+, and we should stop support v10 in the next major anyway.

ronag replied:
I'm good with that

ronag replied:
Removed in separate PR. Ignore it here.

ronag commented on test/agent.js:
```diff
@@ -484,3 +484,16 @@ test('fails with invalid URL', t => {
+test('constructor validations', t => {
+  t.plan(2)
+  t.throw(() => new Agent({ factory: 'ASD' }), InvalidArgumentError, 'throws on invalid opts argument')
+  t.throw(() => new Agent({ maxRedirections: 'ASD' }), InvalidArgumentError, 'throws on invalid opts argument')
```
also try negative integer

ShogunPanda replied:
Good point. On my way!

ShogunPanda replied:
Done!

ronag commented on lib/agent/redirect.js:
```diff
@@ -83,16 +83,38 @@ class RedirectHandler {
+        TLDR: undici always ignores 3xx response bodies.
+
+        Redirection is used to serve the requested resource from another URL, so it is assumes that
+        no body is generated (and thus can be ignored). Even thought generating a body is not prohibited.
```
```suggestion
        no body is generated (and thus can be ignored). Even though generating a body is not prohibited.
```

ShogunPanda replied:
Thanks for noticing. Fixed.

ShogunPanda commented:
@mcollina `-100` was not possible. For now I've added `--lines 95 --branches 95 --functions 95 --statements 95` so we have a good starting point

@ronag I've removed all the unwanted ignores and update this PR description accordingly. I'll leave it to you to write test for the missing cases (or reinstate `ignores` ;)). 

Once this is merged I think also #603 can be merged, isn't it?
